piet-svg/src/text.rs

```diff
@@ -1,114 +1,356 @@
 //! Text functionality for Piet svg backend
 
-use std::ops::RangeBounds;
+use std::{
+    collections::HashSet,
+    fs, io,
+    ops::RangeBounds,
+    sync::{Arc, Mutex},
+};
 
+use font_kit::{
+    handle::Handle,
+    source::{Source, SystemSource},
+    sources::{mem::MemSource, multi::MultiSource},
+};
 use piet::kurbo::{Point, Rect, Size};
 use piet::{
-    Error, FontFamily, HitTestPoint, HitTestPosition, LineMetric, TextAttribute, TextStorage,
+    Color, Error, FontFamily, FontStyle, FontWeight, HitTestPoint, HitTestPosition, LineMetric,
+    TextAlignment, TextAttribute, TextStorage,
 };
+use rustybuzz::{Face, UnicodeBuffer};
 
 type Result<T> = std::result::Result<T, Error>;
 
 /// SVG text (unimplemented)
-#[derive(Debug, Clone)]
-pub struct Text;
+#[derive(Clone)]
+pub struct Text {
+    source: Arc<Mutex<MultiSource>>,
```

Comment:
Is this a `Mutex` only because we're using `Arc`, or are we actually expecting this to be used from multiple threads?

Reply:
I chose the type this way because I wanted the text object to be `Clone`, `Send`, and `Sync`. You can think of it as a shared object that contains all fonts and allows you to fetch them. The fact that this might involve loading from disk is abstracted away.

The alternative to this strategy is using `Rc<RefCell>` and having a separate object per thread. I would think this is wasteful if you're loading fonts from disk multiple times. If you're only going to use the text object from the main thread then this isn't an issue.

What would really be nice is a library like `font-kit` that was a bit more high-level, where there is just 1 object for font loading that reference counts any shared state under the hood. That way you could be more clever about avoiding locks and blocking.

I think I'm rambling, so I'll stop. Basically, I can leave it as is or change it to `Rc<RefCell>`, at your preference. I think I need to use interior mutability, because we expect `RenderContext::Text` to be cheap to clone.

Reply:
This is fine, thanks!